### ISSUE TITLE
LazyLock not working as expected with deref

### ISSUE THREAD

**Issue by Norlock** [rust-lang/rust#140340]
I'm not sure if this is a bug inside uuid or LazyLock, but I believe LazyLock should prevent any changes without the lock being acquired. 

I tried this code:

```rust
<code>
const TA: LazyLock<uuid::Uuid> = LazyLock::new(|| Uuid::new_v4());

pub fn main() {
  info!("{:?} {:?}", *TA, *TA);

}
```

I expected to see this happen: 
Two similar UUID's 

Instead, this happened: 
[2025-04-26T15:38:56Z INFO  splittable_be::templates] 3b9bd2f5-c2e1-433b-962e-1f662acdf5b1 c6258c4c-5bec-4ad5-a5f9-0818ec13537c

### Meta

`rustc --version --verbose`:
```
rustc 1.86.0 (05f9846f8 2025-03-31)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: x86_64-unknown-linux-gnu
release: 1.86.0
LLVM version: 19.1.7
```

</p>
</details>

**Comment by hanna-kruppe** [comment]
It’s because `TA` is a `const`, not a `static`. Constants are (conceptually) duplicated into every place where they’re used, which means every execution of `*TA` creates a new temporary `LazyLock` independent from the others. Clippy warns about this kind of footgun by default. Arguably rustc should also warn.

**Comment by y21** [comment]
> Arguably rustc should also warn.

#132146 adds this as a lint to rustc as well

**Comment by Norlock** [comment]
I don't get why doesn't const have a static lifetime? This makes the static more const than const itself?

**Comment by theemathas** [comment]
A const behaves similarly to if you copy-pasted the code into each point you use the const.

Your code is roughly equivalent to this:
```rust
pub fn main() {
  info!("{:?} {:?}", *LazyLock::new(|| Uuid::new_v4()), *LazyLock::new(|| Uuid::new_v4()));
}
```

**Comment by traviscross** [comment]
Note also that this is independent of interior mutability, and this same behavior can be demonstrated without it.  E.g.:

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&code=%2F%2F+This+demonstrates+how+dereferencing+a+constant+cannot+in+general+be%0A%2F%2F+expected+to+produce+the+same+value+--+how+it+can+even+produce%0A%2F%2F+values+nondeterministically+--+and+how+this+is+independent+of%0A%2F%2F+interior+mutability%2C+which+the+example+below+does+not+use+at+all.%0A%2F%2F%0A%2F%2F+See%3A%0A%2F%2F%0A%2F%2F+-+https%3A%2F%2Fgithub.com%2Frust-lang%2Frust%2Fissues%2F140340%0A%2F%2F%0A%2F%2F+Author%3A+TC%0A%2F%2F+Date%3A+2025-04-28%0A%0Ause+core%3A%3A%7Bmarker%3A%3APhantomData%2C+ops%3A%3ADeref%7D%3B%0A%0Aconst+C%3A+W%3Cu8%3E+%3D+W%3A%3Anew%28%7C%7C+rand%28%29%29%3B%0A%0Afn+main%28%29+%7B%0A++++println%21%28%22%7B%7D+%7B%7D+%7B%7D+%7B%7D%22%2C+%2AC%2C+%2AC%2C+%2AC%2C+%2AC%29%3B%0A++++%2F%2F~%5E++++++~~~~~~~~~~~+These+will+be+four+random+values.%0A%7D%0A%0Apub+struct+W%3CT%2C+F+%3D+fn%28%29+-%3E+T%3E+%7B%0A++++f%3A+F%2C%0A++++_marker%3A+PhantomData%3CT%3E%2C%0A%7D%0A%0Aimpl%3CT%2C+F%3A+Fn%28%29+-%3E+T%3E+W%3CT%2C+F%3E+%7B%0A++++pub+const+fn+new%28f%3A+F%29+-%3E+Self+%7B%0A++++++++Self+%7B+f%2C+_marker%3A+PhantomData+%7D%0A++++%7D%0A%7D%0A%0Aimpl%3CF%3A+Fn%28%29+-%3E+u8%3E+Deref+for+W%3Cu8%2C+F%3E+%7B%0A++++type+Target+%3D+u8%3B%0A++++fn+deref%28%26self%29+-%3E+%26Self%3A%3ATarget+%7B%0A++++++++static+VS%3A+%5Bu8%3B+256%5D+%3D+%7B%0A++++++++++++let+mut+vs+%3D+%5B0u8%3B+256%5D%3B%0A++++++++++++let+mut+i+%3D+0usize%3B%0A++++++++++++while+i+%3C+256+%7B%0A++++++++++++++++vs%5Bi%5D+%3D+i+as+u8%3B%0A++++++++++++++++i+%2B%3D+1%3B%0A++++++++++++%7D%0A++++++++++++vs%0A++++++++%7D%3B%0A++++++++%26VS%5B%28self.f%29%28%29+as+usize%5D%0A++++%7D%0A%7D%0A%0Apub+fn+rand%28%29+-%3E+u8+%7B%0A++++unsafe+extern+%22C%22+%7B%0A++++++++pub+fn+getrandom%28%0A++++++++++++buf%3A+%2Aconst+u8%2C%0A++++++++++++len%3A+usize%2C%0A++++++++++++flags%3A+u32%2C%0A++++++++%29+-%3E+isize%3B%0A++++%7D%0A++++let+mut+buf+%3D+0u8%3B%0A++++unsafe+%7B+assert_eq%21%281%2C+getrandom%28%26raw+mut+buf%2C+1%2C+0%29%29+%7D%3B%0A++++buf%0A%7D%0A)

**Comment by oli-obk** [comment]
jup, I think this is sufficiently tracked by #132146
